### ISSUE TITLE
Expose pub sub API

### ISSUE THREAD

**Issue by zbyte64** [ipfs/py-ipfs-api#114]
Pub sub functionality is now available through the IPFS HTTP API, would like the client to support such functionality.

**Comment by whereswaldon** [comment]
You're right that we should have that! We'll take a stab at it soon unless you're interested in adding it?

**Comment by suzaku** [comment]
@whereswaldon Can I help with this? I'm just starting to learn more about IPFS, but I've written a lot of Python code, so I guess I can do this if you give me some documentation about this API.

**Comment by whereswaldon** [comment]
@suzaku We'd love the help. The HTTP API is documented here: https://ipfs.io/docs/api/#apiv0pubsubls
You may have to scroll down to reach it. Find the heading `api/v0/pubsub/ls`, which is the first HTTP endpoint related to Pub/Sub.
It ought to be possible to use a lot of the existing HTTP-request-related code to do this. Look at how existing functions are implemented as a guide.
Thanks!

**Comment by suzaku** [comment]
@whereswaldon 

Thanks, I'll take a look at the documentation.
By the way, just out of curiosity, why isn't `requests` used directly?

**Comment by whereswaldon** [comment]
There were certain types of request that required more complex management
than others. We actually used to do a lot more in the http logic. We
switched to requests in the last year, I think. Before that, I managed a
bunch of generators that wrote multipart http request bodies. It was a
mess.
Anyway, I think we still need the abstraction that the HTTPClient class
provides. I could be wrong though. If you still think we could use requests
directly after making this change, we can discuss a refactor.

**Comment by suzaku** [comment]
@whereswaldon

Thanks for replying.
The additional methods for `pubsub` API are easy to add, but I can't run the tests on my laptop.
There seems to be some kind of compatibility issue with `ipfs`, for example, I would see the following error:

```python
___________________ IpfsApiTest.test_add_filepattern_from_dirname ____________________

self = <tests.IpfsApiTest testMethod=test_add_filepattern_from_dirname>

    def test_add_filepattern_from_dirname(self):
        res = self.api.add(self.fake_dir, pattern=self.pattern)

        try:
>           assert self._sort_by_key(self.fake_dir_fnpattern_res) == self._sort_by_key(res)
E           AssertionError: assert [{'Hash': 'Qm...test2/fssdf'}] == [{'Hash': 'Qmb...'Size': '22'}]
E             At index 0 diff: {'Name': 'fake_dir', 'Hash': 'QmbPzQruAEFjUU3gQfupns6b8USr8VrD9H71GrqGDXQSxm'} != {'Name': 'fake_dir', 'Hash': 'QmbPzQruAEFjUU3gQfupns6b8USr8VrD9H71GrqGDXQSxm', 'Size': '124'}
E             Full diff:
E             - [{'Hash': 'QmbPzQruAEFjUU3gQfupns6b8USr8VrD9H71GrqGDXQSxm', 'Name': 'fake_dir'},
E             ?                                                            ---------------------
E             + [{'Hash': 'QmbPzQruAEFjUU3gQfupns6b8USr8VrD9H71GrqGDXQSxm',
E             +   'Name': 'fake_dir',
E             +   'Size': '124'},
E             {'Hash': 'QmT5rV6EsKNSW619SntLrkCxbUXXQh4BrKm3JazF2zEgEe',
E             -   'Name': 'fake_dir/test2'},
E             ?                           -
E             +   'Name': 'fake_dir/test2',
E             +   'Size': '73'},
E             {'Hash': 'Qmb1NPqPzdHCMvHRfCkk6TWLcnpGJ71KnafacCMm6TKLcD',
E             -   'Name': 'fake_dir/test2/fssdf'}]
E             ?                                 ^^
E             +   'Name': 'fake_dir/test2/fssdf',
E             ?                                 ^
E             +   'Size': '22'}]

test/functional/tests.py:263: AssertionError
```

which shows that `ipfs` returns an extra field.

**Comment by whereswaldon** [comment]
What ipfs version are you using?

**Comment by suzaku** [comment]
@whereswaldon It's 0.4.13.

**Comment by whereswaldon** [comment]
Hrm. We should future-proof the tests... Should be simple enough to ignore
that key.

**Comment by suzaku** [comment]
OK, so before adding new features, I'll create a new branch to make sure all tests pass in both my development environment and Travis.

**Comment by AdamTheAnalyst** [comment]
+1 for this. Happy to help out if you need a hand

**Comment by whereswaldon** [comment]
@AdamTheAnalyst Perhaps you and @suzaku should coordinate on the current status of this?

**Comment by alexander255** [comment]
No, it's not hard or anything. Just follow the existing pattern in `client.py` and add some tests. We just need somebody to do that and contribute the result.

**Comment by slandau3** [comment]
If no one has any objections, I'd like to give this issue a shot.

**Comment by slandau3** [comment]
PR #123 opened

**Comment by lordcirth** [comment]
I believe #123 resolved this - should this issue be closed?

**Comment by slandau3** [comment]
Yep, this PR was merged awhile back. Please close

**Comment by skliarie** [comment]
Does py-ipfs-http-client 0.4.3 has the pubsub API?
How can I use it? (found nothing on HTTP Client reference page)...

**Comment by alexander255** [comment]
Yes, pubsub is supported. Were you looking looking at the old py-ipfs-api 0.4.3 reference by any chance? If yes, which link did you follow to get there?

For the current reference, see here: https://ipfs.io/ipns/12D3KooWEqnTdgqHnkkwarSrJjeMP2ZJiADWLYADaNvUb6SQNyPF/docs/http_client_ref.html#ipfshttpclient.Client.pubsub

**Comment by skliarie** [comment]
my mistake, it works, thank you for the link
